linkerd/http-route/src/http/filter/redirect.rs

```diff
@@ -0,0 +1,100 @@
+use super::ModifyPath;
+use crate::http::RouteMatch;
+use http::{
+    uri::{Authority, InvalidUri, Scheme, Uri},
+    StatusCode,
+};
+
+#[derive(Clone, Debug, Default, Hash, PartialEq, Eq)]
+pub struct RedirectRequest {
+    pub scheme: Option<Scheme>,
+    pub host: Option<String>,
+    pub port: Option<u16>,
+    pub path: Option<ModifyPath>,
+    pub status: Option<StatusCode>,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum InvalidRedirect {
+    #[error("redirects may only replace the path prefix when a path prefix match applied")]
+    InvalidReplacePrefix,
+
+    #[error("redirect produced an invalid location: {0}")]
+    InvalidLocation(#[from] http::Error),
+
+    #[error("redirect produced an invalid authority: {0}")]
+    InvalidAuthority(#[from] InvalidUri),
+
+    #[error("no authority to redirect to")]
+    MissingAuthority,
+}
+
+#[derive(Clone, Debug)]
+pub struct Redirection {
+    pub status: http::StatusCode,
+    pub location: Uri,
+}
+
+// === impl RedirectRequest ===
+
+impl RedirectRequest {
+    pub fn apply(
+        &self,
+        orig_uri: &http::Uri,
+        rm: &RouteMatch,
+    ) -> Result<Option<Redirection>, InvalidRedirect> {
+        let location = {
+            let scheme = self
+                .scheme
+                .clone()
+                .or_else(|| orig_uri.scheme().cloned())
+                .unwrap_or(Scheme::HTTP);
+
+            let authority: Authority = match (self.host.as_deref(), self.port) {
+                // If a host is configured, use it and whatever port is configured.
+                (Some(h), p) => p
+                    .map(|p| format!("{}:{}", h, p).parse())
+                    .unwrap_or_else(|| h.parse())?,
+                // If a host is NOT configured, use the request's original host and either an
+                // overridden port or the original port.
+                (None, p) => {
+                    let h = orig_uri.host().ok_or(InvalidRedirect::MissingAuthority)?;
+                    p.or_else(|| orig_uri.port_u16())
+                        .map(|p| format!("{}:{}", h, p).parse())
+                        .unwrap_or_else(|| h.parse())?
```

Comment:
also, we may want to use the `TryFrom<String>` impl for `Authority` rather than `FromStr`? the `TryFrom<String>` impl will [try to construct an `Authority` from the `String` converted into a `Bytes`](https://docs.rs/http/latest/src/http/uri/authority.rs.html#454-461) while [the `FromStr` impl calls the `TryFrom<&'a str>` impl](https://docs.rs/http/latest/src/http/uri/authority.rs.html#463-469), which [will copy the `&str`'s bytes into a new `Bytes`](https://docs.rs/http/latest/src/http/uri/authority.rs.html#425-443). since we have ownership of the strings returned by `format!` here, we may want to use the conversion that takes the `String` by value, as it won't allocate a new `Bytes` and copy the string into it, but instead [uses the same allocation](https://docs.rs/bytes/1.1.0/src/bytes/bytes.rs.html#836-840).